python/fastfec.py

```diff
@@ -1,59 +1,386 @@
+
+            with open('file.fec', 'rb') as f:
+                fastfec.parse_as_files(f, open=open_output_file)
+
+        Close the library:
+
+            fastfec.free()
+        """
+
+    def __init__(self):
+        self.__init_lib()
 
         # Initialize
-        self.file_descriptors = {}
-        self.last_filename = None
-        self.last_fd = None
-        self.bytes_written = 0
-        self.buffer_read_fn = BUFFER_READ(
-            make_read_buffer(file_handle))
-        write_callback = self.provide_write_callback(output_directory)
-        self.write_callback_fn = CUSTOM_WRITE(write_callback)
-
-        self.persistent_memory_context = self.libfastfec.newPersistentMemoryContext(
-            1)
-        self.fec_context = self.libfastfec.newFecContext(
-            self.persistent_memory_context, self.buffer_read_fn, BUFFER_SIZE, self.write_callback_fn, BUFFER_SIZE, None, None, None, 0, 1)
-
-    def init_lib(self):
-        # TODO: ensure this works on any platform
-        self.libfastfec = CDLL(os.path.join(SCRIPT_DIR, "../zig-out/lib/libfastfec.dylib"))
-
-        # Lay out arg/res types
+        self.persistent_memory_context = self.libfastfec.newPersistentMemoryContext()
+
+    def parse(self, file_handle):
+        """Parses the input file line-by-line
+
+        Arguments:
+            file_handle -- An input stream for reading a .fec file
+
+        Returns:
+            An iterator that receives the form name and a dictionary
+            object describing each line in the file"""
+        # Set up a queue to be able to yield output
+        queue = Queue()
+        done_processing = object()
+
+        # Provide a custom line callback
+        buffer_read_fn = provide_read_callback(file_handle)
+        line_callback_fn = CUSTOM_LINE(
+            provide_line_callback(queue))
+        fec_context = self.libfastfec.newFecContext(
+            self.persistent_memory_context, buffer_read_fn, BUFFER_SIZE, CUSTOM_WRITE(
+                0),
+            BUFFER_SIZE, line_callback_fn, 0, None, None, None, 0, 1, 0)
+
+        # Run the parsing in a separate thread
+        def task():
+            self.libfastfec.parseFec(fec_context)
+            queue.put(done_processing)
+
+        Thread(target=task, args=()).start()
+
+        # Yield processed lines
+        while True:
+            line = queue.get()
+            if line == done_processing:
+                break
+            yield line
+            queue.task_done()
+
+        # Free FEC context
+        self.libfastfec.freeFecContext(fec_context)
+
+    def parse_as_files(self, file_handle, open_function=open):
+        """Parses the input file into output files
+
+        Arguments:
+            file_handle -- An input stream for reading a .fec file
+            open_function -- An optional function to open an output file for writing. This can be
+                             set to customize the output stream for each parsed .csv file"""
```

Comment:
Can we specify what's returned in the docstring?

Reply:
Done